src/UserSpaceInstrumentation/Trampoline.cpp

```diff
@@ -0,0 +1,155 @@
+  constexpr uint64_t kMax32BitOffset = 0x7fffffff;
+  const uint64_t page_size = sysconf(_SC_PAGE_SIZE);
+
+  // Try to fit an interval of length `size` below `code_range`.
+  auto optional_range_index = LowestIntersectingAddressRange(taken_ranges, code_range);
+  if (!optional_range_index) {
+    return ErrorMessage(absl::StrFormat("code_range %#x-%#x is not in taken_ranges.",
+                                        code_range.start, code_range.end));
+  }
+  while (optional_range_index.value() > 0) {
+    // Place directly to the left of the take interval we are in...
+    uint64_t address_trampoline = taken_ranges[optional_range_index.value()].start - size;
+    // ... but round down to the previous page boundary.
+    address_trampoline = (address_trampoline / page_size) * page_size;
+    AddressRange range_trampoline = {address_trampoline, address_trampoline + size};
+    optional_range_index = LowestIntersectingAddressRange(taken_ranges, range_trampoline);
+    if (!optional_range_index) {
+      // We do not intersect any taken interval. Check if we are close enough to code_range:
+      // code_range is above range_trampoline we will need to jmp back and forth in these ranges
+      // with 32 bit offsets. If no distance is greater than 0x7fffffff this is safe.
+      if (code_range.end - range_trampoline.start <= kMax32BitOffset) {
+        return range_trampoline;
+      }
+      // If we are already beyond the close range there is no need to go any further.
+      break;
+    }
+  }
+
+  // Try to fit an interval of length `size` above `code_range`.
+  optional_range_index = HighestIntersectingAddressRange(taken_ranges, code_range);
+  if (!optional_range_index) {
+    return ErrorMessage(absl::StrFormat("code_range %#x-%#x is not in taken_ranges.",
+                                        code_range.start, code_range.end));
+  }
+  do {
+    const uint64_t address_trampoline =
+        ((taken_ranges[optional_range_index.value()].end + (page_size - 1)) / page_size) *
+        page_size;
+    // Check if we ran out of address space.
+    if (address_trampoline >= 0xffffffffffffffff - size) {
+      break;
+    }
+    AddressRange range_trampoline = {address_trampoline, address_trampoline + size};
+    optional_range_index = HighestIntersectingAddressRange(taken_ranges, range_trampoline);
+    if (!optional_range_index) {
+      // We do not intersect any taken interval. Check if we are close enough to code_range:
+      // code_range is below range_trampoline we will need to jump back and forth in these ranges
+      // with 32 bit offsets. If no distance is greater than 0x7fffffff this is safe.
+      if (range_trampoline.end - code_range.start <= kMax32BitOffset) {
+        return range_trampoline;
+      }
+      // If we are already beyond the close range there is no need to go any further.
+      break;
+    }
+  } while (optional_range_index.value() < taken_ranges.size());
```

Comment:
isn't this always true?

Reply:
Indeed. This was left over from a prvious version. Changed to "true".